File: src/cpp/20_operacje_na_plikach/zad12.cpp
/*
Tytul: Wczytaj i wypisz tresc pliku.
Tresc zadania: Otrzymujesz napis reprezentujacy sciezke do pliku tekstowego.
Wczytaj plik i wypisz jego tresc. Dane wejsciowe: Napis reprezentujacy sciezke
do pliku. Dane wyjsciowe: Napis zawierajacy tresc pliku.

*/
#include <cassert>
#include <experimental/filesystem>
#include <fstream>
#include <iostream>
#include <stdexcept>

namespace filesys = std::experimental::filesystem;

std::string nazwaPliku(const std::string &sciezka) {
  return filesys::path(sciezka).filename();
}

void przeniesPlik(const std::string &sciezka,
                  const std::string &sciezkaDocelowa) {

  try {
    filesys::copy_file(sciezka,
                       sciezkaDocelowa + filesys::path::preferred_separator +
                           nazwaPliku(sciezka),
                       filesys::copy_options::overwrite_existing);
    filesys::remove(sciezka);
  } catch (filesys::filesystem_error &e) {
    std::cout << e.what() << std::endl;
  }
}

std::string znajdzRozszerzenie(std::string sciezka) {

  filesys::path obiekt(sciezka);

  if (obiekt.has_extension())
    return obiekt.extension().string();

  return "";
}

void przeniesPliki(const std::string &sciezka,
                   const std::string &sciezkaDocelowa) {

  for (const auto plik : filesys::directory_iterator(sciezka)) {
    if (znajdzRozszerzenie(plik.path()) == ".csv")
      przeniesPlik(plik.path(), sciezkaDocelowa);
  }
}

void testPrzeniesPliki() {

  // stworz foldery test1 i test2
  filesys::create_directory("test1");
  filesys::create_directory("test2");

  // stworz pliki test1/plik1.csv i test1/plik2.csv

  std::string sciezka1 = "test1/plik1.csv";
  std::string sciezka2 = "test1/plik2.csv";

  std::ofstream plik1(sciezka1);
  plik1 << "test1; test2" << std::endl;
  plik1.close();

  std::ofstream plik2(sciezka2);
  plik2 << "test3; test4" << std::endl;
  plik2.close();

  // przenies pliki z test1 do test2
  przeniesPliki("test1", "test2");

  // sprawdz czy pliki zostaly przeniesione
  assert(filesys::exists("test2/plik1.csv"));
  assert(filesys::exists("test2/plik2.csv"));
  assert(!filesys::exists("test1/plik1.csv"));
  assert(!filesys::exists("test1/plik2.csv"));

  // usun foldery test1 i test2
  filesys::remove_all("test1");
  filesys::remove_all("test2");
}

int main() {

  testPrzeniesPliki();

  return 0;
}

File: src/python/20_operacje_na_plikach/zad12.py
"""
ZAD-12 — Przenieś wszystkie pliki CSV do jednego folderu (rekurencyjnie)

**Poziom:** ★★☆
**Tagi:** `files`, `move`, `csv`, `recursive`

### Treść

Otrzymujesz ścieżkę folderu źródłowego i docelowego. Przenieś wszystkie pliki `.csv` z folderu źródłowego **i wszystkich jego podfolderów** do folderu docelowego.

### Wejście

* 1 linia: `src_folder`
* 2 linia: `dst_folder`

### Wyjście

Brak.

### Przykład

**Wejście:**

```
C:\Users\Username\Projekty
D:\Dane\CSV
```

**Wyjście:**
*(brak)*

### Uwagi

* Jeśli w folderze docelowym istnieje już plik o tej samej nazwie, zadanie wymaga zdefiniowania zachowania (np. zmiana nazwy / pominięcie) — jeśli sprawdzarka tego nie doprecyzowuje, przyjmij jedną spójną strategię w całym rozwiązaniu.

"""

import pathlib


def przenies_pliki_csv(folder_1, folder_2):
    """
    Funkcja przenosi pliki csv z folderu folder_1 do folderu folder_2.
    """
    for plik in pathlib.Path(folder_1).glob("**/*.csv"):
        plik.rename(pathlib.Path(folder_2) / plik.name)


def test_przenies_pliki_csv():

    # stworz foldery testowe
    folder_1 = pathlib.Path("test_1")
    folder_1.mkdir(parents=True, exist_ok=True)
    folder_2 = pathlib.Path("test_2")
    folder_2.mkdir(parents=True, exist_ok=True)

    # stworz pliki testowe
    plik_1 = folder_1 / "plik_1.csv"
    plik_1.touch()
    plik_2 = folder_1 / "plik_2.csv"
    plik_2.touch()

    # przenies pliki
    przenies_pliki_csv(folder_1, folder_2)

    # sprawdz czy pliki zostaly przeniesione
    assert pathlib.Path(folder_2 / "plik_1.csv").exists()
    assert pathlib.Path(folder_2 / "plik_2.csv").exists()
    assert not pathlib.Path(folder_1 / "plik_1.csv").exists()
    assert not pathlib.Path(folder_1 / "plik_2.csv").exists()

    # usun foldery testowe
    import shutil

    shutil.rmtree("test_1")


if __name__ == "__main__":

    test_przenies_pliki_csv()
